### PR TITLE
Bad submission_url in forms from Aggregate fixed

### LINKED ISSUE #873: Bad submission_url in forms from Aggregate should not be loaded
#### Software and hardware versions 
Collect v1.4.16

#### Problem description
A form with a bad submission URL can be loaded into Collect's Fill Blank Form. This is a problem because submission URLs override Collect's server settings so you can never send data to Aggregate after you fill out a blank form.

#### Steps to reproduce the problem
Create a form with a bad URL (e.g., `https:/server.com` instead of `https://server.com`). Fill out form and submit. To reproduce, try these forms:
* [bug.xlsx](https://github.com/opendatakit/collect/files/853365/bug.xlsx)
* [bug.xml.txt](https://github.com/opendatakit/collect/files/853366/bug.xml.txt)

If you'd like to make your own forms, for XLSForm (put a `submission_url` header with the value below in the `settings` sheet). For XForm, 
```
<model>
  <submission action="https:/server.com/submission" method="form-data-post"/>
  <instance>
```

#### Expected behavior
Users should be alerted as soon as possible when the submission URL is bad. 

There is already code in `DiskSyncTask.java` (https://github.com/opendatakit/collect/pull/716) that prevents forms that are loaded from the SD card or from Google Drive to be removed. This issue is for forms that are downloaded from Aggregate.

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Fixes #873

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java [hunk 1]
```diff
@@ -200,12 +202,26 @@ private boolean isXformsListNamespacedElement(Element e) {
                                 manifestUrl = null;
                             }
                             break;
+
+                        case "submission":
+                            submissionUrl = XFormParser.getXMLText(child, true);
+                            if (!UrlUtils.isValidUrl(submissionUrl)) {
+                                submissionUrl = null;
+                    }
```
Comment (yanokwa) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java]: Are you sure this bracket is in the correct place?

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java [hunk 2]
```diff
@@ -144,6 +145,7 @@ private boolean isXformsListNamespacedElement(Element e) {
                 String description = null;
                 String downloadUrl = null;
                 String manifestUrl = null;
+                String submissionUrl = null;
```
Comment (yanokwa) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java]: Forms often have empty submission URLs. If this is set to null, you won't be able to distinguish between an empty submission URL (which is fine) and a bad submission URL. You'll hit line 218 every time you try to parse a blank form before download. You can reproduce this by checking out this branch and trying to download new forms. Setting `submissionUrl` to an empty string is the likely correct fix.

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java [hunk 3]
```diff
@@ -200,12 +202,26 @@ private boolean isXformsListNamespacedElement(Element e) {
                                 manifestUrl = null;
                             }
                             break;
+
```
Comment (yanokwa) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java]: Don't think you need this whitespace

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java]: sure

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java (hunk 2 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java]: Sorry, I completely forgot about that case.

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java (hunk 1 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormListTask.java]: I'll fix this

**Comment by yanokwa** [comment]
I can confirm this code works well if I push the form on the SD card, but it does not work well if I download that same form from Aggregate. 

Try downloading the "Bug" form from https://nafundi-test.appspot.com. Fill out a form and try to submit it. It will crash with the following:

```
AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:325)
AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
AndroidRuntime: 	at java.lang.Thread.run(Thread.java:761)
AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Host name may not be null
AndroidRuntime: 	at org.opendatakit.httpclientandroidlib.util.Args.containsNoBlanks(Args.java:81)
AndroidRuntime: 	at org.opendatakit.httpclientandroidlib.HttpHost.<init>(HttpHost.java:80)
AndroidRuntime: 	at org.opendatakit.httpclientandroidlib.HttpHost.<init>(HttpHost.java:135)
AndroidRuntime: 	at org.odk.collect.android.utilities.WebUtils.enablePreemptiveBasicAuth(WebUtils.java:171)
AndroidRuntime: 	at org.odk.collect.android.tasks.InstanceUploaderTask.uploadOneSubmission(
AndroidRuntime: 	at org.odk.collect.android.tasks.InstanceUploaderTask.processChunk(InstanceUploaderTask.java:530)
AndroidRuntime: 	at org.odk.collect.android.tasks.InstanceUploaderTask.doInBackground(InstanceUploaderTask.java:554)
AndroidRuntime: 	at org.odk.collect.android.tasks.InstanceUploaderTask.doInBackground(InstanceUploaderTask.java:75)
AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:305)
AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
```

This is because the form has a bad URL (note the single slash)
`<submission action="https:/nafundi-test.appspot.com" method="form-data-post"/>`

It looks like the forms.db has the broken URL, so you'll likely need to validate the submission URL again, and I think DownloadFormsTask.java is where that code lives.

**Comment by shobhitagarwal1612** [comment]
Thanks for the help :smiley:

**Comment by shobhitagarwal1612** [comment]
@yanokwa  Please review

**Pull request merged**